app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/AddressController.scala

```diff
@@ -429,4 +445,19 @@ trait AddressController[A <: AddressJourneyType] {
       case _                                 => None
     }
 
+  private def registeredWithId(journey: AddressJourneyType): Option[Boolean] =
+    journey match {
+      case onboarding: AddressJourneyType.Onboarding             =>
+        onboarding match {
+          case Onboarding.RegistrationReadyAddressJourney(_)              => Some(false)
+          case Onboarding.IndividualSupplyingInformationAddressJourney(_) => Some(false)
+          case Onboarding.SubscriptionReadyAddressJourney(_)              => Some(true)
+        }
+      case subscription: AddressJourneyType.ManagingSubscription =>
+        subscription match {
+          case ManagingSubscription.SubscribedAddressJourney(subscribed) =>
+            Some(subscribed.subscribedDetails.registeredWithId)
+        }
+      case _: AddressJourneyType.Returns                         => Some(true)
```

Comment:
this might not be `true` since you can do a return when you've registered without id. You can avoid having to consider whether or not someone regesiterd with or without id where you don't need to though if you take a more localised approach to displaying the exit page. So if you had:
```scala
def ukAddressNotAllowed(): Action[AnyContent] = Action {
  exit_page()
} 
```
in some controller somewhere then inside `RegistrationChangeAddressController.scala` and `RegistrationEnterAddressController.scala` you could just have:
```
override def enterPostcode(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
override def enterPostcodeSubmit(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
override def ukAddress(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
override def ukAddressSubmit(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
```
and in the `SubscribedChangeAddressController.scala` the same thing expect you guard the redirect with a check on whether they registered with or without id.  Then you don't need this method or any changes to the `AddressController` or the `AddresControllerSpec` - you'd only need changes to the specific controllers and their specs. 
Also, having two separate endpoints defined for the same url in the routes file won't work since you'll always be redirect to one controller and the states won't match up in some of the journeys.

Reply:
There's no natural controller for the "global" exit page, so maybe it could just go in the `StartController` with the above solution. I think there's a couple of exit pages there already.

Reply:
I did look at putting it in the `StartController` first but this exit page is not a global one but rather a specific one concerning addresses which is why I placed it under the address the package. That's my interpretation of it. Is that right?

Reply:
> this might not be `true` since you can do a return when you've registered without id. You can avoid having to consider whether or not someone regesiterd with or without id where you don't need to though if you take a more localised approach to displaying the exit page. So if you had:
> 
> ```scala
> def ukAddressNotAllowed(): Action[AnyContent] = Action {
>   exit_page()
> } 
> ```
> 
> in some controller somewhere then inside `RegistrationChangeAddressController.scala` and `RegistrationEnterAddressController.scala` you could just have:
> 
> ```
> override def enterPostcode(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> override def enterPostcodeSubmit(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> override def ukAddress(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> override def ukAddressSubmit(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> ```
> 
> and in the `SubscribedChangeAddressController.scala` the same thing expect you guard the redirect with a check on whether they registered with or without id. Then you don't need this method or any changes to the `AddressController` or the `AddresControllerSpec` - you'd only need changes to the specific controllers and their specs.
> Also, having two separate endpoints defined for the same url in the routes file won't work since you'll always be redirect to one controller and the states won't match up in some of the journeys.

I'll give this some thought. Not sure if I understand it yet.

Reply:
> this might not be `true` since you can do a return when you've registered without id. You can avoid having to consider whether or not someone regesiterd with or without id where you don't need to though if you take a more localised approach to displaying the exit page. So if you had:
> 
> ```scala
> def ukAddressNotAllowed(): Action[AnyContent] = Action {
>   exit_page()
> } 
> ```
> 
> in some controller somewhere then inside `RegistrationChangeAddressController.scala` and `RegistrationEnterAddressController.scala` you could just have:
> 
> ```
> override def enterPostcode(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> override def enterPostcodeSubmit(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> override def ukAddress(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> override def ukAddressSubmit(): Action[AnyContent] = Action { _ => Redirect(ukAddressNotAllowed)}
> ```
> 
> and in the `SubscribedChangeAddressController.scala` the same thing expect you guard the redirect with a check on whether they registered with or without id. Then you don't need this method or any changes to the `AddressController` or the `AddresControllerSpec` - you'd only need changes to the specific controllers and their specs.
> Also, having two separate endpoints defined for the same url in the routes file won't work since you'll always be redirect to one controller and the states won't match up in some of the journeys.

One of the problems with your proposed solution is the back-link. The way I have it at the moment is that the back-link can be easily set inside the respective subtype of the AddressController type. 

What you are proposing is a general handler for this. With that we cannot easily go back to the page the user can from unless we thread through from which page the redirect to the exit-page was triggered.

Reply:
I could pass a URL (string) to the controller method:

```scala
def ukAddressNotAllowed(backLink :String)
```

and then feed that into the general template.

I'll go with that and raise another PR.

Reply:
oh yeah, that's a good point - it's a trade off then between having to cater for all journey types in the `AddressController` trait and being able to determine the back link in some generic exit page . The latter would overall be more annoying to do I think - but for the first one, we need different URL's for the exit page depending on the journey type then. One way could be to  tag on `exit/cannot-use-service-non-resident-uk-address` to the existing urls that exist for the sections like:
```
/registration/create-account/exit/cannot-use-service-non-resident-uk-address
/registration/create-account/exit/cannot-use-service-diferent-non-resident-uk-address
/account-home/manage-account/exit/cannot-use-service-non-resident-uk-address
```
We'd have to get that double checked though

Reply:
that's how it works at the moment - each subtype of the `AddressController` trait has it's own exit url. If you have another look at the router files, I made those changes in there.

Reply:
scratch that, I see what you mean. Yeah we would need to change the urls as you described above. Who can do that? Jean I suppose.

Reply:
If we do make the URL changes, then I think this is good to go. Basically your solution approaches the problem from the opposite angle but with the price that we have to feed the backLink url through. 

We need to make a decision :smile:

Reply:
there's a security risk with having a URL as a query parameter as another URL - I can't remember exactly when but I remember it being raised at some point on another project. What might be easiest then is to go with what you've got, but maybe think about the `Option[Call]` approach in my other comment to reduce the number of cases we have to worry about with this addition

Reply:
the risk was something about people being able to share a url in an email or something which looked like a legitimate tax platform url but make the back link point to something malicious.

Reply:
That's fine with me. I'll review the review comments and make the changes.

Reply:
fixed as agreed